crypto/core_namemap.c

```diff
@@ -8,62 +8,55 @@
  */
 
 #include "internal/namemap.h"
-#include <openssl/lhash.h>
-#include "crypto/lhash.h"      /* ossl_lh_strcasehash */
 #include "internal/tsan_assist.h"
+#include "internal/hashtable.h"
 #include "internal/sizes.h"
 #include "crypto/context.h"
 
-/*-
- * The namenum entry
- * =================
- */
-typedef struct {
-    char *name;
-    int number;
-} NAMENUM_ENTRY;
+#define NAMEMAP_HT_BUCKETS 4096
 
-DEFINE_LHASH_OF_EX(NAMENUM_ENTRY);
+HT_START_KEY_DEFN(namenum_key)
+HT_DEF_KEY_FIELD_CHAR_ARRAY(name, 64)
+HT_END_KEY_DEFN(NAMENUM_KEY)
 
 /*-
  * The namemap itself
  * ==================
  */
 
+typedef char STRING;
+typedef STACK_OF(STRING) NAMES;
+
+DEFINE_STACK_OF(STRING)
+DEFINE_STACK_OF(NAMES)
+
 struct ossl_namemap_st {
     /* Flags */
     unsigned int stored:1; /* If 1, it's stored in a library context */
 
+    HT *namenum_ht;        /* Name->number mapping */
+
     CRYPTO_RWLOCK *lock;
-    LHASH_OF(NAMENUM_ENTRY) *namenum;  /* Name->number mapping */
+    STACK_OF(NAMES) *numnames;
```

Comment:
Why use a stack here for the reverse mapping?  Doing so means you need to maintain the rwlock around this data, when you could just define a new key and datatype for the same hashtable, which should let you eliminate that

Reply:
That wouldn't work. There is one to many mapping for num to name map. Thus you need a list of name entries in each num entry. For that we would have to define a completely new data type with rcu semantics to avoid the rwlock.

However IMO this is not needed as the num to name map is used much less than the mapping other way around.

Reply:
but if you define your key as <NUM>, and map it to an entry with a character array NAME, then you should be fine.  Ostensibly, if we're looking for a name that matches a given number, we don't care which name we get, we just care that we get _a_ name.  Looking at the current implementation of ossl_namemap_num2name:

```
static void do_num2name(const char *name, void *vdata)
{
    struct num2name_data_st *data = vdata;

    if (data->idx > 0)
        data->idx--;
    else if (data->name == NULL)
        data->name = name;
}

const char *ossl_namemap_num2name(const OSSL_NAMEMAP *namemap, int number,
                                  size_t idx)
{
    struct num2name_data_st data;

    data.idx = idx;
    data.name = NULL;
    if (!ossl_namemap_doall_names(namemap, number, do_num2name, &data))
        return NULL;
    return data.name;
}
```

We currently iterate over every entry in the table looking for a matching number at the i'th location (which itself seems silly, because no ordering is guaranteed).  When we hit the i'th index, we fill out the corresponding name, and return that.

Every caller passes 0 as the idx, so we're always looking for the 'first' entry (which again, seems silly since the hash table has no real ordering).  As such, why not:
1) Create a second key for the table, defined as a unit32_t representing the number
2) on insert, attempt to insert a character string matching the name to be added against its allocated number.  If it fails, fine, it means we already have a name assigned to that number. If we succeed, good
3) On num2name lookup, we just do a lookup of the number key, and return whatever name is found, which should be just as good as either a full list traversal, or a stack iteration, only faster, because theres no (or very little) iteration involved.

Reply:
That is not possible. Please note we have the doall names calls so we must store all the names inserted. You cannot choose to store just one name.

Reply:
BTW the stack is very efficient on retrieving n-th value so IMO the hashtable lookup won't be any faster.

Reply:
yes, and we can store all inserted names, but we don't need to store the reverse mapping for all inserted number (which was my point).  We only need to store one number to name mapping, which is why the use of multiple data types is useful here.

regarding the use of a stack, it certainly is efficient for looking up the n'th value of a list, but my point is we don't ever actually need to do that, because (a) every caller of this the num2name function only looks up the 1st entry, because its impossible to know, given our current implementation, what the n'th entry will ever be (i.e. a caller has no way to know definitively what the 1st, 2nd, 3rd, etc number to name mapping is).  Even if the insertion order is known, there is no guarantee that is the order which the hash table will return them in during an iteration (as that will be completely dependent on the hash function used, and the sharding of the table.

Eliminating the use of the stack, and replacing it with a separate data type in the hash table lets you eliminate the use of the rdlock entirely, relying instead on the lockless behavior of the hashtable

Reply:
You need stack of stacks to hold the data. The fast access is useful for the number to stack of strings mapping. We need to store all the names. Although the order is not that important, it is actually good that we keep it because then the first name returned will be the _canonical_ one.

The lock is needed anyway on the write side to keep the mappings consistent and the read side needs it only for the num to name mapping which is not used in critical paths.

Reply:
wait, are you saying that the last name inserted onto the stack is considered the canonical name for a given number?  Isn't that somewhat arbitrary, as it might be impacted by the (equally arbitrary) order in which providers are loaded?

Reply:
> wait, are you saying that the last name inserted onto the stack is considered the canonical name for a given number? Isn't that somewhat arbitrary, as it might be impacted by the (equally arbitrary) order in which providers are loaded?

Not the last name but the first name. Given that our providers are usually loaded first this works fine. Besides we already say that the first name should be considered the canonical name somewhere in the documentation so other providers should follow this naming convention as well. Otherwise names like `2.16.840.1.101.3.4.2.1` would be returned for SHA256 instead of the canonical `SHA2-256` name.